rfcs/convergence/exposing-popper-options.md

```diff
@@ -0,0 +1,224 @@
+These options are spread into the component's props:
+
+```tsx
+<Tooltip position="bottom" align="center" />
+<Popover position="bottom" align="center" />
+<Menu position="bottom" align="center" />
+```
+
+There are other popper options that are used in some components but not others, examples of these options are:
+
+- `coverTarget` - allows the positioned element to cover the target
+
+There are also additional options that are not used by any converged component that might be used in the future like:
+
+- `autoSize` - Allows the positioned element to change size based on viewport size
+- `flipBoundary` - Determines what bounds to use to flip the positioning
+
+Issues and bugs can be triggered because certain use cases require more precise configuration of popper. The most likely
+result would be that a popper configuration is exposed for one component but not others.
+
+## Detailed Design or Proposal
+
+### Equal popper configuration
+
+Each converged component that uses popper should allow an equal level of configuration. This ensures that our positioned
+components can behave consistently across the library. Any fixes that require exposing another popper configuration will
+be available to **all** components that use popper.
+
+This proposal does not mean that all popper options must be public, but that if one component exposes a popper configuration,
+then all components should do so equally.
+
+We can achieve this with an agreement that the current `PositioningProps` would be supported by all v9 components, while
+internal popper options can still belong to `PopperOptions` type in the `react-positioning` package.
+
+### Shorthand `placement` configuration
+
+The next step to this proposal is to make sure that all components which use a positioned element should allow users
+to configure popper through a `placement` prop that behaves like a shorthand. An extra helper will be required to resolve
+shorthand to full popper configuration internally, but the external API for `placement` is very clean for users.
+
+The usage examples below will detail the usage of this shorthand.
+
+```tsx
+// @fluentui/react-positioning
+export type PositioningShorthandValue =
+  | 'above'
+  | 'above-start'
+  | 'above-end'
+  | 'below'
+  | 'below-start'
+  | 'below-end'
+  | 'before'
+  | 'before-top'
+  | 'before-bottom'
+  | 'after'
+  | 'after-top'
+  | 'after-bottom';
+
+export type PositioningShorthand = PositioningProps | PositioningShorthandValue;
+
+export function resolvePopperOptions(shorthand: PositioningShorthand, config: PositioningProps) {
+  if (typeof shorthand === 'object') {
+    return {
+      ...shorthand,
+      ...config,
+    };
+  }
+
+  if (typeof shorthand === 'string') {
+    return {
+      ...parseStringShorthand(shorthand),
+      ...config,
+    };
+  }
+
+  return config;
+}
+
+export function parseStringShorthand(shorthand) {
+  return {
+    position,
+    align,
+  };
+}
```

Comment:
It'd probably be better to let the components do the `{ ...shorthand, ...config }` spread themselves, rather than baking it into `resolvePopperOptions`.
* It doesn't support default values (`config` overrides the `placement` prop).
* Some of the values could be calculated based on the value of the placement prop. For example, Tooltip's `offset` value is `placement.offset + arrowHeight` (add a fixed amount to whatever the user passed in).
* Spreading an object is pretty natural and easy for someone to write if that's what they need:

I'd propose to delete `resolvePopperOptions` and _only_ have the `parseStringShorthand` function (maybe call it `resolvePositioningShorthand`?) to convert from shorthand to object.

Example usage might be:
```ts
const popperOptions = { offset: 4, ...resolvePositioningShorthand(state.placement) };
popperOptions.offset += arrowHeight;
const popper = usePopper(popperOptions);
```

--

```suggestion
export function resolvePositioningShorthand(placement: PositioningShorthand): PositioningShorthand;
export function resolvePositioningShorthand(placement: PositioningShorthand | undefined) : PositioningShorthand | undefined {
  if (typeof placement === 'string) {
    // ... parse the string, or lookup in a table ...
    return {
      position,
      align,
    };
  }
  
  return placement;
}
```

Reply:
I don't mind, parsing shorthand is always gonna be awkward, so it will have to be handled smwhere. More importantly the actual API for consumers will be greatly simplified.

I'm fine with going along with this proposal 👍

Reply:
In any case `usePopper` will continue to only accept a single object, and it **should not** accept shorthand values

Reply:
Also reverted to `positioning` it just fits well with everything currently there too

Reply:
Highly approve `positioning` 👍